### PR TITLE
fixes #9060, #8175 - refactor module and support dispatch router 

### PR DESCRIPTION
**Pull request by stbenjam**
I feel like the module is too Katello-specific.  While not perfectly generic, this makes more things parameterized and sets a better stage for me to add configuration for the dispatcher.

### CODE REVIEW

**Comment by ekohl** [comment]
I don't know qpid, but at first glance the code looks good. I can do a stricter review later, but remind me if I forget :)

File: templates/qpidd.conf.erb [hunk 1]
```diff
@@ -12,27 +12,27 @@
...
-# "qpidd --help" or "man qpidd" for more details. 
-#cluster-mechanism=ANONYMOUS 
-log-enable=error+
-log-to-syslog=yes
-auth=no
-require-encryption=yes
+# (Note: no spaces on either side of '='). Using default settings:
+# "qpidd --help" or "man qpidd" for more details.
+#cluster-mechanism=ANONYMOUS
+log-enable=<%= scope.lookupvar("qpid::log_level") %>
```
Comment (ehelms) [inline templates/qpidd.conf.erb]: What about using `@` for in scope variable lookup?

File: manifests/params.pp [hunk 2]
```diff
@@ -3,13 +3,24 @@
...
+  $ssl_cert_password_file  = undef
+  $ssl_cert_name           = undef
+  $ssl_require_client_auth = true
 
   $user_groups = []
 
   $user = 'qpidd'
   $group = 'qpidd'
 
+
```
Comment (ehelms) [inline manifests/params.pp]: Could probably remove this.

**Comment by ehelms** [comment]
Minor comments but otherwise looks great! Thanks for adding tests with the refactor!

File: templates/qpidd.conf.erb (hunk 1 above)
Comment (stbenjam) [inline templates/qpidd.conf.erb]: It's coming out of qpid::config, so it needs to do the scope lookup - but I didn't know about `scope[]`, neat!

File: templates/qpidd.conf.erb [hunk 3]
```diff
@@ -12,27 +12,27 @@
...
+log-to-syslog=<%= scope.lookupvar("qpid::log_to_syslog") ? 'yes' : 'no' %>
+auth=<%= scope.lookupvar("qpid::auth") ? 'yes' : 'no' %>
 <% if scope.lookupvar("qpid::ssl") -%>
-ssl-require-client-authentication=yes
+require-encryption=yes
```
Comment (bechtoldt) [inline templates/qpidd.conf.erb]: please add a class param for this

**Comment by bechtoldt** [comment]
I added some changes to @stbenjam's repo (https://github.com/stbenjam/puppet-qpid/pull/1) which makes #13 obsolete

**Comment by stbenjam** [comment]
Thanks @bechtoldt Will take it once the rest is settled

**Comment by stbenjam** [comment]
I've changed the way I handle the router to use the concat pattern, as you can have multiple resources of each time in `qdrouter.conf`. 

The outputs of the following puppet configurations were validated by the qpid team:

Configuration for a Katello:
```
class {'qpid::router':
  config_file => '/tmp/qdrouter.conf'
}

# Configure for SSL
qpid::router::ssl_profile {'router-ssl':
  ca   => '/some/where/ca.pem',
  cert => '/some/where/cert.pem',
  key  => '/some/where/cert.key',
}

# Listen for other qpid dispatch routers
qpid::router::listener {'hub':
  addr        => '0.0.0.0',
  port        => 9700,
  role        => 'inter-router',
  ssl_profile => 'router-ssl',
}

# Listen for katello-agent clients
qpid::router::listener {'clients':
  addr        => '0.0.0.0',
  port        => 5671,
  ssl_profile => 'router-ssl',
}

# Connect dispatch router to the local qpid
qpid::router::connector {'broker':
  addr        => '127.0.0.1',
  port        => 9701,
  ssl_profile => 'router-ssl',
  role        => 'on-demand',
}
```

Configuration for an Isolated Capsule:
```
class {'qpid::router':
  config_file => '/tmp/qdrouter.conf'
}

# Configure for SSL
qpid::router::ssl_profile {'router-ssl':
  ca   => '/some/where/ca.pem',
  cert => '/some/where/cert.pem',
  key  => '/some/where/cert.key',
}

# Listen for katello-agent clients
qpid::router::listener {'clients':
  addr        => '0.0.0.0',
  port        => 5671,
  ssl_profile => 'router-ssl',
}

# Connect to Katello dispatch router instance
qpid::router::connector {'hub':
  addr        => 'satellite.example.com',
  port        => 9700,
  ssl_profile => 'router-ssl',
  role        => 'inter-router',
}
```

**Comment by stbenjam** [comment]
@ekohl Do you have any opinions about the concat pattern, and my implementation here?

**Comment by ekohl** [comment]
@stbenjam I think it makes sense and defines are a prime example where the concat pattern makes sense.

File: manifests/router/connector.pp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+#
+define qpid::router::connector(
+  $addr        = '0.0.0.0',
+  $port        = 5672,
+  $sasl_mech   = 'ANONYMOUS',
+  $role        = '',
+  $ssl_profile = '',
+){
+
+  concat_fragment {"qdrouter+connector_${title}.conf":
```
Comment (ekohl) [inline manifests/router/connector.pp]: I thought you should use $name inside the deinfe.

File: manifests/router.pp [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+class qpid::router(
+  $router_id               = $qpid::router::params::router_id,
+  $mode                    = $qpid::router::params::mode,
+  $config_file             = $qpid::router::params::config_file,
+  $container_name          = $qpid::router::params::container_name,
+) inherits qpid::router::params {
+
+  class { 'qpid::router::install': } ~>
+  class { 'qpid::router::config': }
+
```
Comment (ekohl) [inline manifests/router.pp]: There's no qpid service that should be notified when the config changes?

File: manifests/router.pp [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+# $config_file::    Dispatch router config file location
+#
+class qpid::router(
+  $router_id               = $qpid::router::params::router_id,
+  $mode                    = $qpid::router::params::mode,
+  $config_file             = $qpid::router::params::config_file,
+  $container_name          = $qpid::router::params::container_name,
+) inherits qpid::router::params {
+
+  class { 'qpid::router::install': } ~>
```
Comment (ekohl) [inline manifests/router.pp]: Generally install doesn't need to refresh the config.

**Comment by ehelms** [comment]
Tested via https://github.com/Katello/katello-deploy/pull/115 and got success.  So, ACK!

**Comment by stbenjam** [comment]
Thanks!

**Pull request merged**